pkg/controllers/state/statenodepool.go

```diff
@@ -0,0 +1,187 @@
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package state
+
+import (
+	"sync"
+	"sync/atomic"
+
+	"github.com/samber/lo"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	v1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+)
+
+// Tracks vended limits for a NodePool that are in-flight
+type NodePoolLimit struct {
+	ReservedNodeLimit atomic.Int64
+}
+
+type NodeClaimState struct {
+	Running  sets.Set[string]
+	Deleting sets.Set[string]
+}
+
+// StateNodePool is a cached version of a NodePool in the cluster that maintains state which is expensive to compute every time it's needed.
+type NodePoolState struct {
+	mu sync.RWMutex
+
+	nodePoolNameToNodeClaimState map[string]NodeClaimState // node pool name -> node claim state (Running and Deleting node claim names)
+	nodeClaimNameToNodePoolName  map[string]string         // node claim name -> node pool name
+	nodePoolNameToNodePoolLimit  map[string]*NodePoolLimit // node pool -> nodepool limit
+
+}
+
+func NewNodePoolState() *NodePoolState {
+	return &NodePoolState{
+		nodePoolNameToNodeClaimState: map[string]NodeClaimState{},
+		nodeClaimNameToNodePoolName:  map[string]string{},
+		nodePoolNameToNodePoolLimit:  map[string]*NodePoolLimit{},
+	}
+}
+
+// Helper methods that hold the lock
+
+// Sets up the NodePoolState to track NodeClaim
+func (n *NodePoolState) SetNodeClaimMapping(npName, ncName string) {
+	if npName == "" || ncName == "" {
+		return
+	}
+	n.mu.Lock()
+	defer n.mu.Unlock()
+
+	n.ensureNodePoolEntry(npName)
+	n.nodeClaimNameToNodePoolName[ncName] = npName
+}
+
+// ReleaseNodePoolNodeLimit releases the NodePoolsNodeLimit
+func (n *NodePoolState) ReleaseNodePoolNodeLimit(npName string) {
+	n.mu.Lock()
+	defer n.mu.Unlock()
+
+	currentlyReserved := n.nodePoolNameToNodePoolLimit[npName].ReservedNodeLimit.Load()
```

Comment:
IMO, you should be able to pass a numeric value for how many you want to release here and the functions should be called something like Add() and Subtract()